src/ReverseProxy/Service/HealthChecks/ActiveHealthCheckMonitor.cs

```
@@ -0,0 +1,181 @@
+            if (activeHealthCheckOptions.Enabled)
+            {
+                _scheduler.ChangePeriod(cluster, activeHealthCheckOptions.Interval ?? _monitorOptions.DefaultInterval);
+            }
+            else
+            {
+                _scheduler.UnscheduleEntity(cluster);
+            }
+        }
+
+        public void OnClusterRemoved(ClusterInfo cluster)
+        {
+            _scheduler.UnscheduleEntity(cluster);
+        }
+
+        public void Dispose()
+        {
+            _scheduler.Dispose();
+        }
+
+        private async Task ProbeCluster(ClusterInfo cluster)
+        {
+            var clusterConfig = cluster.Config;
+            if (!clusterConfig.HealthCheckOptions.Active.Enabled)
+            {
+                return;
+            }
+
+            // Policy must always be present if the active health check is enabled for a cluster.
+            // It's validated and ensured by a configuration validator.
+            var policy = _policies.GetRequiredServiceById(clusterConfig.HealthCheckOptions.Active.Policy);
+            var allDestinations = cluster.DynamicState.AllDestinations;
+            var probeTasks = new List<(Task<HttpResponseMessage> Task, CancellationTokenSource Cts)>(allDestinations.Count);
+            try
+            {
+                foreach (var destination in allDestinations)
+                {
+                    var timeout = clusterConfig.HealthCheckOptions.Active.Timeout ?? _monitorOptions.DefaultTimeout;
+                    var cts = new CancellationTokenSource(timeout);
+                    try
+                    {
+                        var request = _probingRequestFactory.CreateRequest(clusterConfig, destination.Config);
+                        probeTasks.Add((clusterConfig.HttpClient.SendAsync(request, cts.Token), cts));
+                    }
+                    catch
+                    {
+                        cts.Cancel();
```

Why are we cancelling here at all?

---

Shouldn't we be capturing the exception here as well?

---

Exceptions thrown as part of probes creation are handled by [try/catch in ForceCheckAll](https://github.com/microsoft/reverse-proxy/blob/2042a88bc9cbc46e8ea536723739bc0134d52a24/src/ReverseProxy/Service/HealthChecks/ActiveHealthCheckMonitor.cs#L59). However, if `IProbingRequestFactory` is broken then we cannot rely on active health check results at all, so there is no point in trying to suppress them for part of destinations to allow other probes being sent because it will anyways lead to an inconsistent system state.

---

`Cancel` call is made to avoid accidentally holding request-related resources in `SendAsync` which get released when a request completes or gets terminated.

---

ProbeCluster is called from 2 places and we're ignoring exceptions in both of those places today.

---

We are not ignoring them, logging is simply not yet implemented. It's [on my plan](https://github.com/microsoft/reverse-proxy/pull/459#issuecomment-705091700) and will be done before converting this from a draft to a real PR.

---

Got it.